### PR TITLE
My stab at floating IP support

### PR DESCRIPTION
**Pull request by tyler-ball**
I updated the `attach_ip` calls to use the correct parameters.  AWS was the only one who wanted an `allocation_id` so I made that specific to AWS.  I _think_ the only other consumer of floating IPs is Openstack (according to Fog), so the rest of the code I tested with Openstack.

I would love some more testing on this - my openstack instance is local and it is... flaky :)

I also added some minor bug fixes like:
* Moved the Mutex to a class variable because `||=` is not thread safe
* Pulled in the changes from https://github.com/chef/chef-provisioning-fog/pull/47 to make IP assignment Idempotent
  * IP removal is not supported yet - TODO

Replaces https://github.com/chef/chef-provisioning-fog/pull/69
Replaces https://github.com/chef/chef-provisioning-fog/pull/47
Fixes https://github.com/chef/chef-provisioning-fog/issues/77
Fixes https://github.com/chef/chef-provisioning-fog/issues/76
\cc @jkeiser @ryancragun @thommay @vills

### CODE REVIEW

File: lib/chef/provisioning/fog_driver/driver.rb [hunk 1]
```diff
@@ -424,46 +425,63 @@ def wait_for_transport(action_handler, machine_spec, machine_options, server)
...
     end
 
     # Attach IP to machine from IP pool
     # Code taken from kitchen-openstack driver
-    #    https://github.com/test-kitchen/kitchen-openstack/blob/master/lib/kitchen/driver/openstack.rb#L196-L207
+    #    https://github.com/test-kitchen/kitchen-openstack/blob/master/lib/kitchen/driver/openstack.rb
     def attach_ip_from_pool(server, pool)
-      @ip_pool_lock ||= Mutex.new
-      @ip_pool_lock.synchronize do
+      @@ip_pool_lock.synchronize do
```
Comment (jkeiser) [inline lib/chef/provisioning/fog_driver/driver.rb]: Why the change to a class variable?  Not necessarily bad, but it could be less performant than it has to be if you are provisioning two separate openstacks at once.

File: lib/chef/provisioning/fog_driver/driver.rb [hunk 2]
```diff
@@ -424,46 +425,63 @@ def wait_for_transport(action_handler, machine_spec, machine_options, server)
...
-      # TODO this is not particularly idempotent. OK, it is not idempotent AT ALL.  Fix.
-      if option_for(machine_options, :floating_ip_pool)
+      pool = option_for(machine_options, :floating_ip_pool)
+      floating_ip = option_for(machine_options, :floating_ip)
+      if pool
         Chef::Log.info 'Attaching IP from pool'
-        action_handler.perform_action "attach floating IP from #{option_for(machine_options, :floating_ip_pool)} pool" do
-          attach_ip_from_pool(server, option_for(machine_options, :floating_ip_pool))
+        allocated_addresses = server.addresses[pool]
+        unless allocated_addresses.nil?
```
Comment (jkeiser) [inline lib/chef/provisioning/fog_driver/driver.rb]: Since you are doing an `else`, can you do `if allocated_address.nil?`?  I think `unless / else` can be hard to flip your mind through.  `if !allocated_address.nil?` is fine too :)

File: lib/chef/provisioning/fog_driver/driver.rb [hunk 3]
```diff
@@ -424,46 +425,63 @@ def wait_for_transport(action_handler, machine_spec, machine_options, server)
...
+          floating_ips = allocated_addresses.select {|a_info| a_info['OS-EXT-IPS:type']=='floating'}
+          if floating_ips.length>0
```
Comment (jkeiser) [inline lib/chef/provisioning/fog_driver/driver.rb]: What happens if `floating_ips.length == 0`?  Shouldn't we attach an ip from the pool?  Might look like this:

```ruby
allocated_addresses = server.addresses[pool] || []
floating_ips = allocated_addresses.select {|a_info| a_info['OS-EXT-IPS:type']=='floating'}
if floating_ips.size > 0
  Chef::Log.info "Server already assigned floating_ips `#{floating_ips}` from pool `#{pool}`"
else
  action_handler.perform_action "Attaching floating IP from pool `#{pool}`" do
    attach_ip_from_pool(server, pool)
  end
end
```

File: lib/chef/provisioning/fog_driver/driver.rb (hunk 1 above)
Comment (jkeiser) [inline lib/chef/provisioning/fog_driver/driver.rb]: On second thought, the class variable is significantly safer, and how often are we going to do the above thing anyway?

File: lib/chef/provisioning/fog_driver/driver.rb [hunk 4]
```diff
@@ -424,46 +425,63 @@ def wait_for_transport(action_handler, machine_spec, machine_options, server)
...
+      @@ip_pool_lock.synchronize do
         Chef::Log.info "Attaching floating IP from <#{pool}> pool"
-        free_addrs = compute.addresses.collect do |i|
-          i.ip if i.fixed_ip.nil? and i.instance_id.nil? and i.pool == pool
+        free_addrs = compute.addresses.map do |i|
+          i.ip if i.fixed_ip.nil? && i.instance_id.nil? && i.pool == pool
         end.compact
         if free_addrs.empty?
-          raise ActionFailed, "No available IPs in pool <#{pool}>"
+          raise RuntimeError, "No available IPs in pool <#{pool}>"
```
Comment (jkeiser) [inline lib/chef/provisioning/fog_driver/driver.rb]: This looks deliberate: what's the reasoning behind the error class change?

**Comment by jkeiser** [comment]
Many questions, but :+1: otherwise, yay!

File: lib/chef/provisioning/fog_driver/driver.rb (hunk 1 above)
Comment (tyler-ball) [inline lib/chef/provisioning/fog_driver/driver.rb]: Yeah, I just stole that from the most recent knife-openstack driver, and my newfound knowledge that `||=` is not thread safe

**Comment by Sliim** [comment]
Hi!
I tested this change on my OpenStack instance! I confirm this fixes #77.

But I have two others issues about floating ips:

- The driver attempt to attach ip but the server is not fully created (I got a 400 error with message: 'No nw_info cache associated with instance'). I think you should wait for server ready before attaching an ip to the server (kitchen-openstack do this: https://github.com/test-kitchen/kitchen-openstack/blob/master/lib/kitchen/driver/openstack.rb#L77-L80).

- The driver is waiting for transport to be connectable but fails.. Stacktrace:
```
- waiting for myserver (d32b3b42-a7bf-482c-93d1-4e0d086b3183 on fog:OpenStack:http://myopenstack:5000/v2.0/tokens) to be connectable (transport up and running) ...[2015-02-24T15:12:14+01:00] WARN: Machine myserver (d32b3b42-a7bf-482c-93d1-4e0d086b3183 on fog:OpenStack:http://myopenstack:5000/v2.0/tokens) was started but SSH did not come up.  Rebooting machine in an attempt to unstick it ...
```
And I get a 404 error.. I'm investigating on this to understand what is the real cause (my OpenStack or this driver)

Thx!

File: lib/chef/provisioning/fog_driver/driver.rb (hunk 2 above)
Comment (tyler-ball) [inline lib/chef/provisioning/fog_driver/driver.rb]: Done

File: lib/chef/provisioning/fog_driver/driver.rb (hunk 3 above)
Comment (tyler-ball) [inline lib/chef/provisioning/fog_driver/driver.rb]: You are right - I made this update.

Is it possible to be assigned an IP from the pool without it being floating:question:

File: lib/chef/provisioning/fog_driver/driver.rb (hunk 4 above)
Comment (tyler-ball) [inline lib/chef/provisioning/fog_driver/driver.rb]: I was getting a `NameError` error because it didn't know about the ActionFailed class.  It looks like that was originally copied from knife-openstack?

**Comment by tyler-ball** [comment]
@Sliim I updated to move floating ip assignment until after the `server.ready?` call.

I am also seeing the SSH transport errors - going to troubleshoot that this afternoon.

File: lib/chef/provisioning/fog_driver/driver.rb (hunk 3 above)
Comment (jkeiser) [inline lib/chef/provisioning/fog_driver/driver.rb]: I don't know offhand, but if it's impossible, then we shouldn't do `allocated_addresses.select {|a_info| a_info['OS-EXT-IPS:type']=='floating'}`, methinks.  We'd also have to be sure the AWS API won't ever return us an empty array (instead of `nil`) for `allocated_addresses`.

File: lib/chef/provisioning/fog_driver/driver.rb (hunk 4 above)
Comment (jkeiser) [inline lib/chef/provisioning/fog_driver/driver.rb]: Sounds good :)  I think if you don't have a specific error class you want to return, you can just raise the string itself.

**Comment by jkeiser** [comment]
All good except the one comment: I suspect there might be a bug lurking in the `if addresses.size > 0` code.

**Comment by tyler-ball** [comment]
@jkeiser I agree, but I'm having trouble testing Openstack - I registered for trystack.org but their service isn't working.  And my local instance won't let me assign a floating ip (actively troubleshooting that one).

**Comment by Sliim** [comment]
@tyler-ball your update fix the IP assignement ;)
I think the SSH transport error is due to bad username configuration in my recipe, here is my code:
```
with_driver 'fog:OpenStack', compute_options: {
  openstack_username: ENV['OS_USERNAME'],
  openstack_api_key: ENV['OS_PASSWORD'],
  openstack_auth_url: "#{ENV['OS_AUTH_URL']}/tokens",
  openstack_tenant: ENV['OS_TENANT_NAME'] }

machine 'beef' do
  action [:ready, :converge]
  machine_options bootstrap_options: {
    flavor_ref: '2',
    key_name: 'pentestlab',
    image_ref: 'b8417aac-e565-4984-a3af-1343484ff3b9', # debian wheezy
    security_groups: ['default'],
    sudo:  true,
    is_windows: false
  }, floating_ip_pool: 'public',
     ssh_username: 'stack'

  recipe 'beef'
end
```
With this recipe, I have not the SSH transport error, I was used `ssh_user` option instead of `ssh_username` and I think it was the cause of this error..

But now I have another error, the driver attempt to assign a new floating IP when the server is connectable, and I get a 400 error because the server already have a floating IP.

Maybe my recipe is not perfect yet, so if you have some advises for me.. :)

For informations, I use the OpenStack Icehouse installed with `Packstack`

**Comment by tyler-ball** [comment]
@Sliim Can you try putting a `require 'pry'; binding.pry` statement at https://github.com/chef/chef-provisioning-fog/blob/tball/floating-ip/lib/chef/provisioning/fog_driver/driver.rb#L438 and seeing what happens as you step through the code?  That if statement is supposed to prevent trying to assign a floating IP address to an instance which already has a floating IP.

**Comment by Sliim** [comment]
Hi @tyler-ball 

It seems there is a mistake for `server.addresses` array values, I also put `binding.pry` statement in `attach_ip` function and look what happen:
```
    - waiting for beef (13fab876-df07-43ce-a60f-56457dd4cc51 on fog:OpenStack:http://172.21.50.10:5000/v2.0/tokens) to be ready ...
    - beef is now ready
From: /home/sliim/codz/pentest/pentestlab/pentestlab-provisioning/vendor/chef-provisioning-fog/lib/chef/provisioning/fog_driver/driver.rb @ line 441 Chef::Provisioning::FogDriver::Driver#attach_floating_ips:

    436:         Chef::Log.info 'Attaching IP from pool'
    437:         allocated_addresses = server.addresses[pool] || []
    438:
    439:         floating_ips = allocated_addresses.select {|a_info| a_info['OS-EXT-IPS:type']=='floating'}
    440:
 => 441:         require 'pry'; binding.pry
    442:
    443:         if floating_ips.size > 0
    444:           Chef::Log.info "Server already assigned floating_ips `#{floating_ips}` from pool `#{pool}`"
    445:         elsif
    446:           action_handler.perform_action "Attaching floating IP from pool `#{pool}`" do

[1] pry(#<Chef::Provisioning::FogDriver::Providers::OpenStack>)> server.addresses
=> {"private-net"=>[{"OS-EXT-IPS-MAC:mac_addr"=>"fa:16:3e:bb:63:19", "version"=>4, "addr"=>"192.168.0.73", "OS-EXT-IPS:type"=>"fixed"}]}
[2] pry(#<Chef::Provisioning::FogDriver::Providers::OpenStack>)> exit

From: /home/sliim/codz/pentest/pentestlab/pentestlab-provisioning/vendor/chef-provisioning-fog/lib/chef/provisioning/fog_driver/driver.rb @ line 492 Chef::Provisioning::FogDriver::Driver#attach_ip:

    488: def attach_ip(server, ip)
    489:   Chef::Log.info "Attaching floating IP <#{ip}>"
    490:   server.associate_address ip
    491:   (server.addresses['public'] ||= []) << { 'version' => 4, 'addr' => ip }
 => 492:   require 'pry'; binding.pry
    493: end

[1] pry(#<Chef::Provisioning::FogDriver::Providers::OpenStack>)> server.addresses
=> {"private-net"=>[{"OS-EXT-IPS-MAC:mac_addr"=>"fa:16:3e:bb:63:19", "version"=>4, "addr"=>"192.168.0.73", "OS-EXT-IPS:type"=>"fixed"}],
 "public"=>[{"version"=>4, "addr"=>"172.21.50.91"}]}
 [2] pry(#<Chef::Provisioning::FogDriver::Providers::OpenStack>)> exit

    - Attaching floating IP from pool `public`#<Chef::Provisioning::Machine::UnixMachine:0x007f69b2ed18b8>

  * machine[beef] action converge
  From: /home/sliim/codz/pentest/pentestlab/pentestlab-provisioning/vendor/chef-provisioning-fog/lib/chef/provisioning/fog_driver/driver.rb @ line 441 Chef::Provisioning::FogDriver::Driver#attach_floating_ips:

    436:         Chef::Log.info 'Attaching IP from pool'
    437:         allocated_addresses = server.addresses[pool] || []
    438:
    439:         floating_ips = allocated_addresses.select {|a_info| a_info['OS-EXT-IPS:type']=='floating'}
    440:
 => 441:         require 'pry'; binding.pry
    442:
    443:         if floating_ips.size > 0
    444:           Chef::Log.info "Server already assigned floating_ips `#{floating_ips}` from pool `#{pool}`"
    445:         elsif
    446:           action_handler.perform_action "Attaching floating IP from pool `#{pool}`" do

[1] pry(#<Chef::Provisioning::FogDriver::Providers::OpenStack>)> server.addresses
=> {"private-net"=>
  [{"OS-EXT-IPS-MAC:mac_addr"=>"fa:16:3e:bb:63:19", "version"=>4, "addr"=>"192.168.0.73", "OS-EXT-IPS:type"=>"fixed"},
     {"OS-EXT-IPS-MAC:mac_addr"=>"fa:16:3e:bb:63:19", "version"=>4, "addr"=>"172.21.50.91", "OS-EXT-IPS:type"=>"floating"}]}
[2] pry(#<Chef::Provisioning::FogDriver::Providers::OpenStack>)> allocated_addresses
=> []
[3] pry(#<Chef::Provisioning::FogDriver::Providers::OpenStack>)>
```
We can see that public address is moved in `private-net` array, and this cause that`allocated_address` is already empty in `attach_floating_ips`.

**Comment by Sliim** [comment]
```
> nova list
+--------------------------------------+------+--------+------------+-------------+----------------------------------------+
| ID                                   | Name | Status | Task State | Power State | Networks                               |
+--------------------------------------+------+--------+------------+-------------+----------------------------------------+
| e3f7b11f-881a-4823-bfc7-4746e546ef6b | beef | ACTIVE | -          | Running     | private-net=192.168.0.74, 172.21.50.91 |
+--------------------------------------+------+--------+------------+-------------+----------------------------------------+
```

**Comment by Sliim** [comment]
Hi!

Here is a quick change that fix the issue:
```
diff --git a/lib/chef/provisioning/fog_driver/driver.rb b/lib/chef/provisioning/fog_driver/driver.rb
index 6682b93..83a9c32 100644
--- a/lib/chef/provisioning/fog_driver/driver.rb
+++ b/lib/chef/provisioning/fog_driver/driver.rb
@@ -434,7 +434,13 @@ module FogDriver
       if pool
 
         Chef::Log.info 'Attaching IP from pool'
-        allocated_addresses = server.addresses[pool] || []
+        allocated_addresses = []
+        server.addresses.each do |k, v|
+          v.each do | addr |
+            allocated_addresses.push addr
+          end
+        end
+
         floating_ips = allocated_addresses.select {|a_info| a_info['OS-EXT-IPS:type']=='floating'}
         if floating_ips.size > 0
           Chef::Log.info "Server already assigned floating_ips `#{floating_ips}` from pool `#{pool}`"
```

Maybe it's not the proper way to do this...

**Comment by tyler-ball** [comment]
@Sliim Thanks for the great debugging information!  Your patch looks valid to me - I don't know what the keys in `server.addresses` are used for.

I'm trying to get setup with a cloud provider to validate your steps.  I'm also looking https://github.com/chef/knife-openstack/blob/ea6afff5e65733dda8c4c69a23d647c4a6d90b8b/lib/chef/knife/openstack_server_create.rb for an idea of how to attach floating IPs after server creation.

Thanks again!

**Comment by tyler-ball** [comment]
Okay, this is ready for final review.  I added logic to remove any currently assigned floating IPs if none is specified on the resource.

**Comment by Sliim** [comment]
Hi @tyler-ball,
just pulled your last changes and it works for me :)

File: lib/chef/provisioning/fog_driver/driver.rb [hunk 5]
```diff
@@ -424,46 +426,82 @@ def wait_for_transport(action_handler, machine_spec, machine_options, server)
       end
     end
 
-    def attach_floating_ips(action_handler, machine_spec, machine_options, server)
-      # TODO this is not particularly idempotent. OK, it is not idempotent AT ALL.  Fix.
-      if option_for(machine_options, :floating_ip_pool)
-        Chef::Log.info 'Attaching IP from pool'
-        action_handler.perform_action "attach floating IP from #{option_for(machine_options, :floating_ip_pool)} pool" do
-          attach_ip_from_pool(server, option_for(machine_options, :floating_ip_pool))
+    def converge_floating_ips(action_handler, machine_spec, machine_options, server)
```
Comment (jkeiser) [inline lib/chef/provisioning/fog_driver/driver.rb]: This is nice and concise, as a function, without being lossfully compact.  I like :)

**Comment by jkeiser** [comment]
:+1:

**Comment by jkeiser** [comment]
Merged as c8df0ed ... do me a favor and give this a shot again? :)

**Comment by cmluciano** [comment]
@tyler-ball I'm having trouble getting this to work with params similar to @Sliim. Is floating_ip_pool => 'public' all that is necessary?

here is my config
```
machine 'cool_machine' do
  recipe 'test_machine'
  machine_options :bootstrap_options => {
    :flavor_ref => '2',
    :image_ref => '7b649906-aa39-439a-b69d-e8955d2621ae',
    :user_id => 'cloud-user',
    :tenant_id => 'test',
    security_groups: ['default'],
    :key_name => 'test',
    :floating_ip_pool => 'public'
  }
end
```

**Comment by tyler-ball** [comment]
@cmluciano That _should_ be all thats required to get an ip assigned from the pool.  What behavior are you getting?

**Comment by cmluciano** [comment]
@tyler-ball it continues to use the private IP even after deleting in openstack and recreating

I gisted the output here https://gist.github.com/cmluciano/4a64e07a6a828fbe4d9d

**Comment by cmluciano** [comment]
Then is just eventually times out

**Comment by tyler-ball** [comment]
That warning is getting raised from https://github.com/chef/chef-provisioning-fog/blob/c8df0ed605dbb2cf80ddb7c384a79378d3b6d5ab/lib/chef/provisioning/fog_driver/driver.rb#L676 - it looks like assigning a floating IP doesn't change `server.public_ip_address`.  @cmluciano can you try putting a `require 'pry'; binding.pry` statement right before that warning is raised?  If you can, try a `server.reload` and see if that makes `server.public_ip_address` return a valid value.  If that doesn't work, we'll have to inspect the server if it doesn't have a public_ip_address to see if we can find the floating IP.

In any case, this seems like a valid issue and should be moved to a new bug.  Will you create a new issue, reference your comment from here and tag me on the issue?
